Allow for persisting the /home/user/ directory

AObuchow opened the pull request:
### What does this PR do?
Allows for persisting the `/home/user/` directory in workspaces that have persistent storage enabled (i.e. they're using the per-user/common or per-workspace PVC strategies). Each workspace will have their own unique `/home/user/` directory (i.e. they do not share `/home/user/` directories).

This feature is disabled by default, but can be enabled with the `config.workspace.persistUserHome.enabled: <bool>` DWOC field. 

When enabled, a volume named `persistentHome` will be added to the DevWorkspace, and mounted to all container components.  If a DevWorkspace already has a volume that mounts to `/home/user/`, the DevWorkspace spec remains unchanged
.
### What issues does this PR fix or reference?
Fix #1097

### Is it tested? How?

#### Normal usage
1. Install DWO using the changes from this PR
2. Configure `config.workspace.persistUserHome.enabled: true` in the DWOC with `kubectl edit dwoc -n $NAMESPACE`:

```diff
apiVersion: controller.devfile.io/v1alpha1  
config:  
 routing:  
   defaultRoutingClass: basic  
 workspace:  
   imagePullPolicy: Always  
+   persistUserHome:  
+     enabled: true  
kind: DevWorkspaceOperatorConfig
```

3. Apply a DevWorkspace that uses persistent storage (i.e. it uses the per-user/common or per-workspace storage strategy. The  [`plain.yaml`](https://github.com/devfile/devworkspace-operator/blob/main/samples/plain.yaml) devworkspace will work for this case (the common/per-user PVC strategy should be used).

4. Once the workspace starts up, verify that the devworkspace's pod spec has a volumeMount that uses the correct PVC & mounts to `/home/user/`

```diff
containers:
    - resources:
        limits:
          memory: 512Mi
        requests:
          memory: 64M
      terminationMessagePath: /dev/termination-log
      name: web-terminal
      command:
        - tail
        - '-f'
        - /dev/null
(...)
      volumeMounts:
        - name: claim-devworkspace
          mountPath: /tmp/workspace-storage
          subPath: workspacecf48085b48de41fa
        - name: claim-devworkspace
          mountPath: /projects
          subPath: workspacecf48085b48de41fa/projects
+        - name: claim-devworkspace
+          mountPath: /home/user/
+          subPath: workspacecf48085b48de41fa/home/user/
        - name: workspace-metadata
          readOnly: true
          mountPath: /devworkspace-metadata
        - name: kube-api-access-7gxvp
          readOnly: true
          mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      terminationMessagePolicy: File
      image: 'quay.io/wto/web-terminal-tooling:next'
```

5. Exec with a shell into the workspace pod and create a file in `/home/user/`, e.g.  `echo "test" > /home/user/myfile.txt`
6. Stop the workspace by doing a `kubectl edit dw plain-devworkspace -n $NAMESPACE` and setting `spec.started` to `false`. Wait for the workspace deployment to scale to 0 pods
7. Start the workspace again by doing a `kubectl edit dw plain-devworkspace -n $NAMESPACE` and setting `spec.started` to `true`. Wait for the workspace to be ready again by checking it's status with `kubectl get dw -n $NAMESPACE -w`
8. Ensure the file you created in step 5. has persisted by doing a `cat /home/user/myfile.txt`

#### Feature override case (/home/user/ volume already defined)
1. Install DWO using the changes from this PR
2. Configure `config.workspace.persistUserHome.enabled: true` in the DWOC
3. Apply a DevWorkspace that already defines a volume that mounts to `/home/user/`, e.g.:

```YAML
kind: DevWorkspace  
apiVersion: workspace.devfile.io/v1alpha2  
metadata:  
 name: plain-dw-with-home-vm  
spec:  
 started: true  
 routingClass: 'basic'  
 template:  
   components:  
     - name: pre-defined-home  
       volume: {}  
     - name: web-terminal  
       container:  
         image: quay.io/wto/web-terminal-tooling:next  
         memoryLimit: 512Mi  
         mountSources: true  
         volumeMounts:  
           - name: pre-defined-home  
             path: /home/user/  
         command:  
          - "tail"  
          - "-f"  
          - "/dev/null"
```
4. Once the devworkspace is deployed, ensure the `persistentHome` volumeMount does not exist, and that the` pre-defined-home` volumeMount exists instead.

### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

AObuchow commented:
/retest

amisevsk commented on pkg/provision/storage/shared.go:
```diff
@@ -277,3 +277,32 @@ func checkPVCTerminating(name, namespace string, api sync.ClusterAPI) (bool, err
 	return pvc.DeletionTimestamp != nil, nil
 }
+
+func shouldPersistHomeDirectory(workspace *common.DevWorkspaceWithConfig) bool {
+	return workspace.Config.Workspace.PersistUserHome != nil && workspace.Config.Workspace.PersistUserHome.Enabled != nil && *workspace.Config.Workspace.PersistUserHome.Enabled
+}
```
By using a default value, we can assume this will never be nil -- it's not possible to unset this value since it should get replaced by the default config.

amisevsk commented on pkg/provision/storage/commonStorage.go:
```diff
@@ -144,6 +154,12 @@ func (p *CommonStorageProvisioner) rewriteContainerVolumeMounts(workspaceId, pvc
 		additionalVolumes[additionalVolume.Name] = true
 	}
 
+	// When we persist the home directory, a volumeMount is created that references the PVC volume.
+	// This volumeMount should not be rewritten
+	if shouldPersistHomeDirectory(workspace) {
+		additionalVolumes[pvcName] = true
+	}
+
```
Could you explain what this is doing and why it's necessary? Seems like a steep cost to have to pass in the full `common.DevWorkspaceWithConfig` rather than just a DevWorkspaceTemplateSpec just for this.

This reads like a misuse of the `additionalVolumes` struct -- it will work because we're adding a container that references the PVC volume directly, but in general we assume that containers aren't directly referring to the PVC anyways. I worry that we're breaking the general flow of steps in `ProvisionStorage` by passing data to `rewriteContainerVolumeMounts` that's already half processed and expecting the function to detect and handle that.

The purpose behind `additionalVolumes` is to hold volumes mounted to the workspace container that _aren't_ controlled by DWO's storage provisioners (i.e. are not the per-user/per-workspace PVC) -- it's strange to optionally add the per-user/per-workspace PVC to this list in that context.

amisevsk commented on pkg/provision/storage/commonStorage.go:
We might need to also consider how we handle workspaces that do not normally need storage when persistsUserHome is enabled -- currently, `ProvisionStorage` bails out early if the workspace does not have any volumes and `mountSources` is false for all containers (i.e. we would create no non-ephemeral PVCs for the workspace). Do we still want to add `/home/user` in that case?

AObuchow replied:
Thanks to the new approach of this PR (that you suggested) this is already handled, since the persistent home devfile volume will be added to the DevWorkspace, resulting in a PVC being provisioned.

amisevsk commented on pkg/provision/storage/shared.go:
```diff
@@ -277,3 +277,32 @@ func checkPVCTerminating(name, namespace string, api sync.ClusterAPI) (bool, err
+func addHomeVolumeMount(workspace *common.DevWorkspaceWithConfig, pvcName string, podAdditions *v1alpha1.PodAdditions) error {
+	homeUserDirPath := "/home/user/"
+	homeVolumeMount := corev1.VolumeMount{
+		Name:      pvcName,
+		MountPath: homeUserDirPath,
+		SubPath:   fmt.Sprintf("%s/home/user/", workspace.Status.DevWorkspaceId),
```
I've two concerns here:

* Defining the home mount in this way puts how volumeMounts are defined in two places -- one is in `(provisioner).rewriteVolumeMounts` and one is now here. StorageProvisioners are no longer have full control over how volumeMounts are handled
* This is using a different subpath scheme for per-workspace PVCs, which normally just use `volumeMount.Name` as the subpath (since they don't need to scope by workspace ID)

AObuchow replied:
> This is using a different subpath scheme for per-workspace PVCs, which normally just use volumeMount.Name as the subpath (since they don't need to scope by workspace ID)

Glad you pointed this out. I figured this would be an okay "oddity" in exchange for less code duplication. Will address this.

amisevsk reviewed (commented):
This is a good first shot at persistent home directories and should work, but there is some messiness I think we should clean up before merging. It feels like the added functionality is being grafted on in places where it's not a perfect fit

A (totally separate) approach maybe worth considering: 

Conceptually, we can think of the effect of `persistUserHome.enabled = true` on a devfile/DevWorkspace level rather than a pods/containers level. We can imagine the setting internally converting our devfiles by adding a volume component and mounting it to all container components:
```diff
 kind: DevWorkspace
 apiVersion: workspace.devfile.io/v1alpha2
 metadata:
   name: plain-devworkspace
 spec:
   template:
     components:
       - name: web-terminal
         container:
           image: quay.io/wto/web-terminal-tooling:next

           memoryLimit: 512Mi
           mountSources: true
+          volumeMounts:
+            - name: persistentHome
+              path: /home/user/
           command:
             - "tail"
             - "-f"
             - "/dev/null"
+      - name: persistentHome
+        volume: {}
```

This is a fairly simple operation, so why don't we just do that, out of step with (i.e. prior) to calling the storage provisioners?

* If persistent home is enabled for the workspace,
  * Check preconditions (maybe don't do anything if user already has mounts added for home?)
  * Edit the flattened (or unflattened?) devfile by adding a volume component and updating all regular components to mount it at `/home/user`

With this approach, we don't need to worry about how the volume is handled, where it's mounted, etc., as the StorageProvisioner code is set up to handle mounting volume components to devfile components already. This would mean we can just plug into the reconcile loop as another (small) step and leave the provisioners code unchanged (and it also means the StorageProvisioners don't need to think about this setting)

AObuchow commented:
@amisevsk +1 for your suggested approach - I agree, the way I currently have this implemented seems a bit like an abuse of the existing storage provisioner code. Currently reworking this PR, and will address your feedback further :)

AObuchow commented on pkg/provision/storage/shared.go:
```diff
@@ -277,3 +277,32 @@ func checkPVCTerminating(name, namespace string, api sync.ClusterAPI) (bool, err
+			if podAdditions.Containers[cIdx].VolumeMounts[vmIdx].MountPath == homeUserDirPath {
+				containerName := podAdditions.Containers[cIdx].Name
+				volumeMountName := podAdditions.Containers[cIdx].VolumeMounts[vmIdx].Name
+				return fmt.Errorf("volumeMount '%s' for container '%s' already mounts to %s", volumeMountName, containerName, homeUserDirPath)
```
> I'm kind of partial to no. 2 above -- if someone wants to configure it themself we shouldn't necessarily mess with that.

Agreed, I ended up changing things so that if a user's devfile already has a volume mounting to `/home/user`, the persistentHome volume is not added and the feature is essentially disabled for that workspace.

AObuchow commented:
@amisevsk I've finally updated this PR according to your feedback and also updated the PR description.

It should be good to review however I realized I can further improve things by adding a warning to the DevWorkspace if persistUserHome is enabled and the Devworkspace already mounts a volume to /user/home/. Will work on adding another commit for this today.

AObuchow commented:
@amisevsk Added the warning condition when the DW already mounts a volume to /home/user/.

Good for review now, sorry for the last-minute changes.

amisevsk commented on pkg/provision/workspace/tokens.go:
```diff
@@ -140,11 +140,17 @@ func checkSATokenPathCollisions(saTokens []v1alpha1.ServiceAccountToken, mountPa
 			for _, saToken := range collidingSATokens {
 				problemNames = append(problemNames, saToken.Name)
 			}
+			sort.Slice(problemNames, func(i, j int) bool {
+				return problemNames[i] < problemNames[j]
+			})
 
 			if len(problemPaths) == 1 {
 				return fmt.Errorf("the following ServiceAccount tokens have the same path (%s) and mount path (%s): %s", path, mountPath, strings.Join(problemNames, ", "))
 			}
 		}
+		sort.Slice(problemNames, func(i, j int) bool {
+			return problemNames[i] < problemNames[j]
+		})
```
It's not clear why this change is here (and why we sort twice, once before checking if the list has >1 items)

AObuchow replied:
> It's not clear why this change is here
While working on this PR, the SA tokens test for when SA tokens have colliding mount paths failed due to the order of the SA tokens in the actual error message not matching the order in the expected output/error message.

I re-ran the CI and the test passed, which made me suspect there was some flakiness in the ordering of the SA tokens.

 (and why we sort twice, once before checking if the list has >1 items)

I made a mistake. My intention was to ensure the SA tokens are always sorted before we return, not before checking if the list has >1 items.

amisevsk commented on pkg/provision/workspace/persistentHome.go:
```diff
@@ -0,0 +1,87 @@
+// limitations under the License.
+//
+
+package workspace
```
Minor nit: this function is more of a library function than a provision function -- we expect packages in `pkg/provision` to sync resources with the cluster, whereas `pkg/library` packages operate on DevWorkspaces themselves.

amisevsk replied:
I'd also consider not using dwerrors here out of principle, as it suggests a `retryError` may be returned which is not the case (we're just reusing it for warnings -- instead I'd return something like `(warning string, err error)`) but I'm not super firm on this one so it's up to you.

Structurally, _maybe_ also consider taking a "return a modified copy" over editing the object passed in -- we do both but IMO it's cleaner to return a copy that you have apply as we do in the flatten package.

AObuchow replied:
> Structurally, _maybe_ also consider taking a "return a modified copy" over editing the object passed in -- we do both but IMO it's cleaner to return a copy that you have apply as we do in the flatten package.

I reworked things to go with your suggested approach.

One thing I went back-and-fourth on while working on this was trying to hide the complexity of the different cases that can occur from the devworkspace controller code. With the latest revision, the controller code is relatively simple, but the return values of `AddPersistentHomeVolume()` are relatively complex (`bool, DWTemplateSpec and error`).

`AddPersistentHomeVolume()` could be simplified by making the controller call `needsPersistentHomeDirectory()` and deal with the different cases itself, but this made the controller code look a bit messy IMO. I'm open to going with this approach, however.

amisevsk commented on pkg/provision/workspace/persistentHome.go:
```diff
@@ -0,0 +1,87 @@
+			if volumeMount.Path == constants.HomeUserDirectory {
+				// If a volume is already being mounted to /home/user/, it takes precedence
+				// over the DWO-provisioned home directory volume.
+				return false, &dwerrors.WarningError{
+					Message: fmt.Sprintf("Unable to add persistentHome volume: volume '%s' already mounts to %s", volumeMount.Name, constants.HomeUserDirectory),
+				}
```
I wonder if it's worth adding the warning here -- the user is doing nothing wrong and cannot really fix the issue themselves, as it's an administrator-level setting. 

This suggests that folks shouldn't manually configure home directories, which I'm not sure is the case (e.g. if your devfile _needs_ persistent home and you want it to work regardless of whether the config is set).

Perhaps a rewording to be more of an "info" warning than a "warning" warning?

AObuchow replied:
> I wonder if it's worth adding the warning here -- the user is doing nothing wrong and cannot really fix the issue themselves, as it's an administrator-level setting. 

Agreed.

I ended up trying to reword the warning, but it might be better to not add a warning to the workspace and instead log this at the controller level (basically, trying to let the admin know what's going on rather than warning the user).

Let me know your thoughts.

amisevsk commented on pkg/provision/workspace/persistentHome.go:
```diff
@@ -0,0 +1,87 @@
+		return nil
+	}
+
+	homeVolume := v1alpha2.Component{
+		Name: constants.HomeVolumeName,
+		ComponentUnion: v1alpha2.ComponentUnion{
+			Volume: &v1alpha2.VolumeComponent{},
+		},
+	}
+	homeVolumeMount := v1alpha2.VolumeMount{
+		Name: constants.HomeVolumeName,
+		Path: constants.HomeUserDirectory,
+	}
+
+	workspace.Spec.Template.Components = append(workspace.Spec.Template.Components, homeVolume)
+	for _, component := range workspace.Spec.Template.Components {
+		if component.Container == nil {
+			continue
+		}
+		component.Container.VolumeMounts = append(component.Container.VolumeMounts, homeVolumeMount)
+	}
```
Does it make sense to do an extra-careful look before we leap here? E.g. apply the changes to a copy of `components`, check that the resulting workspace is still valid via `devfilevalidation.ValidateComponents(components)` and only return the added home if it doesn't break the workspace?

My main concern here is applying a change that takes a valid devfile and makes it invalid through some edge case (e.g. if the user has a volume named `persistentHome` that mounts somewhere else).

AObuchow replied:
+1 good call, I added the use of `devfilevalidation.ValidateComponents(components)` and added an extra test for the case where the user already has a volume named `persistentHome`

amisevsk reviewed (commented):
Generally looking good :+1:

Some minor organizational feedback

AObuchow commented on controllers/workspace/devworkspace_controller.go:
```diff
@@ -289,6 +290,13 @@ func (r *DevWorkspaceReconciler) Reconcile(ctx context.Context, req ctrl.Request
+	} else if shouldPersistHome {
+		workspace.Spec.Template = *workspaceWithHomeVolume
+	}
+
```
Agreed, that's a lot clearer, and returning a bool to decide what to do with the result feels incorrect. 

I went back-and-fourth on this, but I decided to remove the workspace warning when the user is already mounting a volume to `/home/user/`, as the user isn't doing anything wrong, and it simplifies things a bit.

AObuchow commented on pkg/provision/workspace/tokens.go:
```diff
@@ -140,11 +140,12 @@ func checkSATokenPathCollisions(saTokens []v1alpha1.ServiceAccountToken, mountPa
 			for _, saToken := range collidingSATokens {
 				problemNames = append(problemNames, saToken.Name)
 			}
-
 			if len(problemPaths) == 1 {
+				sort.Strings(problemNames)
 				return fmt.Errorf("the following ServiceAccount tokens have the same path (%s) and mount path (%s): %s", path, mountPath, strings.Join(problemNames, ", "))
 			}
 		}
+		sort.Strings(problemNames)
 		return fmt.Errorf("multiple ServiceAccount tokens share the same path and mount path: %s", strings.Join(problemNames, ", "))
```
Agreed. I've only encountered the CI tests failing once from the issue this is trying to fix.
We could just leave it out and see if it pops up in the future? :man_shrugging: 

Alternatively, the SA tokens test could be reworked a bit to prevent this flaky test (would involve using a regex to split up the error message into SA tokens, and then sort that result).

AObuchow replied:
Sounds good :)

AObuchow commented:
Squashed the fixups and just updated the commit message for https://github.com/devfile/devworkspace-operator/pull/1105/commits/d637ab89213e8d8cc09034d47cea460e63776261 to remove the mention of a workspace warning being applied when a volume already mounts to `/home/user/`
